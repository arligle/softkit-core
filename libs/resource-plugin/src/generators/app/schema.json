{
  "$schema": "http://json-schema.org/schema",
  "$id": "SaasBuildKitApplicationGenerator",
  "title": "SaasBuildKit Application Options Schema",
  "description": "SaasBuildKit Application Options Schema.",
  "cli": "nx",
  "type": "object",
  "properties": {
    "name": {
      "description": "The name of the application.",
      "type": "string",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What name would you like to use for the node application?"
    },
    "directory": {
      "description": "The directory of the new application.",
      "type": "string"
    },
    "lintCommandName": {
      "type": "string",
      "default": "lint",
      "description": "The name of lint nx command in project.json"
    },
    "skipFormat": {
      "description": "Skip formatting files.",
      "type": "boolean",
      "default": false,
      "x-priority": "internal"
    },
    "skipPackageJson": {
      "description": "Do not add dependencies to `package.json`.",
      "type": "boolean",
      "default": false,
      "x-priority": "internal"
    },
    "linter": {
      "description": "The tool to use for running lint checks.",
      "type": "string",
      "enum": [
        "eslint",
        "none"
      ],
      "default": "eslint"
    },
    "unitTestRunner": {
      "description": "Test runner to use for unit tests.",
      "type": "string",
      "enum": [
        "jest",
        "none"
      ],
      "default": "jest"
    },
    "e2eTestRunner": {
      "type": "string",
      "enum": [
        "jest",
        "none"
      ],
      "description": "Test runner to use for end to end (e2e) tests",
      "default": "none"
    },
    "tags": {
      "description": "Add tags to the application (used for linting).",
      "type": "string"
    },
    "frontendProject": {
      "description": "Frontend project that needs to access this application. This sets up proxy configuration.",
      "type": "string",
      "x-priority": "important"
    },
    "setParserOptionsProject": {
      "type": "boolean",
      "description": "Whether or not to configure the ESLint `parserOptions.project` option. We do not do this by default for lint performance reasons.",
      "default": false
    },
    "strict": {
      "type": "boolean",
      "description": "Adds strictNullChecks, noImplicitAny, strictBindCallApply, forceConsistentCasingInFileNames and noFallthroughCasesInSwitch to tsconfig.",
      "default": true
    },
    "i18n": {
      "description": "Add support for i18n.",
      "type": "boolean",
      "default": true
    },
    "health": {
      "description": "Setup health check controller.",
      "type": "boolean",
      "default": true
    },
    "title": {
      "description": "The title of the application.",
      "type": "string",
      "$default": {
        "$source": "argv",
        "index": 0
      }
    },
    "description": {
      "description": "The description of the application.",
      "type": "string",
      "$default": {
        "$source": "argv",
        "index": 0
      }
    },
    "auth": {
      "description": "Configure authentication.",
      "type": "boolean",
      "default": true
    },
    "db": {
      "description": "Configure database.",
      "type": "boolean",
      "default": true
    },
    "dbType": {
      "description": "Database type.",
      "type": "string",
      "enum": [
        "postgres"
      ]
    },
    "dbName": {
      "description": "Database name.",
      "type": "string",
      "$default": {
        "$source": "argv",
        "index": 0
      }
    },
    "appPort": {
      "description": "Application port.",
      "type": "number",
      "default": 9000,
      "minimum": 1025,
      "maximum": 65535
    },
    "contactName": {
      "description": "Contact person name.",
      "type": "string",
      "default": "Unknown"
    },
    "contactEmail": {
      "description": "Contact person email.",
      "type": "string",
      "default": "john.doe@example.com"
    },
    "contactUrl": {
      "description": "Contact url or website.",
      "type": "string",
      "default": "https://example.com/"
    },
    "testsFolderName": {
      "description": "Tests folder name.",
      "type": "string",
      "default": "__tests__"
    },
    "configureJestConfig": {
      "description": "Generates a tailored Jest configuration file for the application when set to true.",
      "type": "boolean",
      "default": true
    },
    "languages": {
      "description": "Set languages for i18n.",
      "type": "array",
      "alias": "l",
      "default": [
        "en"
      ],
      "x-prompt": {
        "type": "list",
        "multiselect": true,
        "message": "What languages would you like to use for i18n? It also possible to specify a region",
        "items": [
          {
            "value": "af",
            "label": "Afrikaans"
          },
          {
            "value": "am",
            "label": "Amharic"
          },
          {
            "value": "ar",
            "label": "Arabic"
          },
          {
            "value": "arn",
            "label": "Mapuche"
          },
          {
            "value": "as",
            "label": "Assamese"
          },
          {
            "value": "az",
            "label": "Azerbaijani"
          },
          {
            "value": "ba",
            "label": "Bashkir"
          },
          {
            "value": "be",
            "label": "Belarusian"
          },
          {
            "value": "bg",
            "label": "Bulgarian"
          },
          {
            "value": "bn",
            "label": "Bangla"
          },
          {
            "value": "bo",
            "label": "Tibetan"
          },
          {
            "value": "br",
            "label": "Breton"
          },
          {
            "value": "bs",
            "label": "Bosnian"
          },
          {
            "value": "ca",
            "label": "Valencian"
          },
          {
            "value": "co",
            "label": "Corsican"
          },
          {
            "value": "cs",
            "label": "Czech"
          },
          {
            "value": "cy",
            "label": "Welsh"
          },
          {
            "value": "da",
            "label": "Danish"
          },
          {
            "value": "de",
            "label": "German"
          },
          {
            "value": "dsb",
            "label": "Lower Sorbian"
          },
          {
            "value": "dv",
            "label": "Maldivian"
          },
          {
            "value": "el",
            "label": "Modern Greek (1453-)"
          },
          {
            "value": "en",
            "label": "English"
          },
          {
            "value": "es",
            "label": "Castilian"
          },
          {
            "value": "et",
            "label": "Estonian"
          },
          {
            "value": "eu",
            "label": "Basque"
          },
          {
            "value": "fa",
            "label": "Persian"
          },
          {
            "value": "fi",
            "label": "Finnish"
          },
          {
            "value": "fil",
            "label": "Pilipino"
          },
          {
            "value": "fo",
            "label": "Faroese"
          },
          {
            "value": "fr",
            "label": "French"
          },
          {
            "value": "fy",
            "label": "Western Frisian"
          },
          {
            "value": "ga",
            "label": "Irish"
          },
          {
            "value": "gd",
            "label": "Gaelic"
          },
          {
            "value": "gl",
            "label": "Galician"
          },
          {
            "value": "gsw",
            "label": "Alsatian"
          },
          {
            "value": "gu",
            "label": "Gujarati"
          },
          {
            "value": "ha",
            "label": "Hausa"
          },
          {
            "value": "he",
            "label": "Hebrew"
          },
          {
            "value": "hi",
            "label": "Hindi"
          },
          {
            "value": "hr",
            "label": "Croatian"
          },
          {
            "value": "hsb",
            "label": "Upper Sorbian"
          },
          {
            "value": "hu",
            "label": "Hungarian"
          },
          {
            "value": "hy",
            "label": "Armenian (see also hyw)"
          },
          {
            "value": "id",
            "label": "Indonesian"
          },
          {
            "value": "ig",
            "label": "Igbo"
          },
          {
            "value": "ii",
            "label": "Nuosu"
          },
          {
            "value": "is",
            "label": "Icelandic"
          },
          {
            "value": "it",
            "label": "Italian"
          },
          {
            "value": "iu",
            "label": "Inuktitut"
          },
          {
            "value": "ja",
            "label": "Japanese"
          },
          {
            "value": "ka",
            "label": "Georgian"
          },
          {
            "value": "kk",
            "label": "Kazakh"
          },
          {
            "value": "kl",
            "label": "Greenlandic"
          },
          {
            "value": "km",
            "label": "Central Khmer"
          },
          {
            "value": "kn",
            "label": "Kannada"
          },
          {
            "value": "kok",
            "label": "Konkani (macrolanguage)"
          },
          {
            "value": "ko",
            "label": "Korean"
          },
          {
            "value": "ky",
            "label": "Kyrgyz"
          },
          {
            "value": "lb",
            "label": "Letzeburgesch"
          },
          {
            "value": "lo",
            "label": "Lao"
          },
          {
            "value": "lt",
            "label": "Lithuanian"
          },
          {
            "value": "lv",
            "label": "Latvian"
          },
          {
            "value": "mi",
            "label": "Maori"
          },
          {
            "value": "mk",
            "label": "Macedonian"
          },
          {
            "value": "ml",
            "label": "Malayalam"
          },
          {
            "value": "mn",
            "label": "Mongolian"
          },
          {
            "value": "moh",
            "label": "Mohawk"
          },
          {
            "value": "mr",
            "label": "Marathi"
          },
          {
            "value": "ms",
            "label": "Malay (macrolanguage)"
          },
          {
            "value": "mt",
            "label": "Maltese"
          },
          {
            "value": "nb",
            "label": "Norwegian Bokmål"
          },
          {
            "value": "ne",
            "label": "Nepali (macrolanguage)"
          },
          {
            "value": "nl",
            "label": "Flemish"
          },
          {
            "value": "nn",
            "label": "Norwegian Nynorsk"
          },
          {
            "value": "no",
            "label": "Norwegian"
          },
          {
            "value": "nso",
            "label": "Sepedi"
          },
          {
            "value": "oc",
            "label": "Occitan (post 1500)"
          },
          {
            "value": "or",
            "label": "Odia (macrolanguage)"
          },
          {
            "value": "pa",
            "label": "Punjabi"
          },
          {
            "value": "pl",
            "label": "Polish"
          },
          {
            "value": "prs",
            "label": "Afghan Persian"
          },
          {
            "value": "ps",
            "label": "Pashto"
          },
          {
            "value": "pt",
            "label": "Portuguese"
          },
          {
            "value": "quz",
            "label": "Cusco Quechua"
          },
          {
            "value": "rm",
            "label": "Romansh"
          },
          {
            "value": "ro",
            "label": "Moldovan"
          },
          {
            "value": "ru",
            "label": "Russian"
          },
          {
            "value": "rw",
            "label": "Kinyarwanda"
          },
          {
            "value": "sah",
            "label": "Yakut"
          },
          {
            "value": "sa",
            "label": "Sanskrit"
          },
          {
            "value": "se",
            "label": "Northern Sami"
          },
          {
            "value": "si",
            "label": "Sinhalese"
          },
          {
            "value": "sk",
            "label": "Slovak"
          },
          {
            "value": "sl",
            "label": "Slovenian"
          },
          {
            "value": "sma",
            "label": "Southern Sami"
          },
          {
            "value": "smj",
            "label": "Lule Sami"
          },
          {
            "value": "smn",
            "label": "Inari Sami"
          },
          {
            "value": "sms",
            "label": "Skolt Sami"
          },
          {
            "value": "sq",
            "label": "Albanian"
          },
          {
            "value": "sr",
            "label": "Serbian (see cnr for Montenegrin)"
          },
          {
            "value": "sv",
            "label": "Swedish"
          },
          {
            "value": "sw",
            "label": "Swahili (macrolanguage)"
          },
          {
            "value": "syr",
            "label": "Syriac"
          },
          {
            "value": "ta",
            "label": "Tamil"
          },
          {
            "value": "te",
            "label": "Telugu"
          },
          {
            "value": "tg",
            "label": "Tajik"
          },
          {
            "value": "th",
            "label": "Thai"
          },
          {
            "value": "tk",
            "label": "Turkmen"
          },
          {
            "value": "tlh",
            "label": "tlhIngan Hol"
          },
          {
            "value": "tn",
            "label": "Tswana"
          },
          {
            "value": "tr",
            "label": "Turkish"
          },
          {
            "value": "tt",
            "label": "Tatar"
          },
          {
            "value": "tzm",
            "label": "Central Atlas Tamazight"
          },
          {
            "value": "ug",
            "label": "Uyghur"
          },
          {
            "value": "uk",
            "label": "Ukrainian"
          },
          {
            "value": "ur",
            "label": "Urdu"
          },
          {
            "value": "uz",
            "label": "Uzbek"
          },
          {
            "value": "vi",
            "label": "Vietnamese"
          },
          {
            "value": "wo",
            "label": "Wolof"
          },
          {
            "value": "xh",
            "label": "Xhosa"
          },
          {
            "value": "yo",
            "label": "Yoruba"
          },
          {
            "value": "zh",
            "label": "Chinese"
          },
          {
            "value": "zu",
            "label": "Zulu"
          }
        ]
      }
    }
  },
  "additionalProperties": false,
  "required": [
    "name"
  ]
}