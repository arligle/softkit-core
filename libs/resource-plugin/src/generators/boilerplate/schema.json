{
  "$schema": "http://json-schema.org/schema",
  "$id": "Boilerplate",
  "title": "",
  "type": "object",
  "properties": {
    "appFolder": {
      "type": "string",
      "description": "The folder name where the boilerplate will be created",
      "x-prompt": "What folder name would you like to use for the boilerplate?"
    },
    "companyName": {
      "type": "string",
      "description": "The name of the company associated with this boilerplate",
      "x-prompt": "Enter the company name to customize the 'name' field in package.json:"
    },
    "tag": {
      "type": "string",
      "default": "latest",
      "description": "The git tag to clone. Defaults to the latest tag if not specified",
      "x-prompt": "Enter the git tag to clone:"
    },
    "repository": {
      "type": "string",
      "default": "https://github.com/softkitit/softkit-nestjs-boilerplate.git",
      "description": "The URL of the git repository to be cloned",
      "x-prompt": "Enter the URL of the git repository to clone:"
    }
  },
  "required": [
    "appFolder",
    "companyName"
  ]
}
