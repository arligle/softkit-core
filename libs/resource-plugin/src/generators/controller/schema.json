{
  "$schema": "http://json-schema.org/schema",
  "$id": "Controller",
  "title": "",
  "type": "object",
  "properties": {
    "projectName": {
      "type": "string",
      "description": "Application name in your workspace, e.g. 'platform'",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What is app name?",
      "x-dropdown": "projects",
      "minLength": 1
    },
    "controllerName": {
      "type": "string",
      "description": "Controller name, usually match with entity name",
      "x-prompt": "What is controller name? (lowercase) e.g. user-group",
      "minLength": 1
    },
    "basePath": {
      "type": "string",
      "description": "App base path name",
      "x-prompt": "What is app base path name?  e.g. api/platform",
      "minLength": 1
    },
    "serviceName": {
      "type": "string",
      "description": "Service name, usually match with entity name and controller name. It's default service that will be injected to controller.",
      "x-prompt": "What is service name? e.g user-group",
      "minLength": 1
    },
    "entityName": {
      "type": "string",
      "description": "Entity name, usually match with controller and service name. It will be used for base DTO generations",
      "x-prompt": "What is entity name? e.g user-group",
      "minLength": 1
    },
    "groupName": {
      "type": "string",
      "description": "Group name is folder name that you want to generate your entity in it e.g. 'users'",
      "x-prompt": "What is group name for generation? It's folder name under the base controllers folder",
      "minLength": 1
    },
    "tenantBaseEntity": {
      "type": "boolean",
      "default": true,
      "description": "It's indicating that this controller is tenant base",
      "x-prompt": "Is this a tenant base controller? (y/n). (Useful for DTO generation, to exclude tenant id from DTOs)"
    },
    "entityIncludesIdField": {
      "type": "boolean",
      "default": true,
      "description": "Indicates if the entity has an 'id' field. If yes, then it will be excluded from the generated DTOs.",
      "x-prompt": "Does your entity include an 'id' field that should be excluded from DTOs? (y/n)"
    },
    "entityIncludesVersionField": {
      "type": "boolean",
      "default": true,
      "description": "Indicates if the entity has a 'version' field. If yes, then it will be excluded from the generated DTOs.",
      "x-prompt": "Does your entity include a 'version' field that should be excluded from DTOs? (y/n)"
    },
    "lintCommandName": {
      "type": "string",
      "default": "lint",
      "description": "The name of lint nx command in project.json"
    }
  },
  "required": [
    "projectName",
    "controllerName",
    "serviceName",
    "entityName",
    "groupName",
    "tenantBaseEntity"
  ]
}