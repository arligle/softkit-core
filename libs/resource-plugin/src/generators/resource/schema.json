{
  "$schema": "http://json-schema.org/schema",
  "$id": "Resource",
  "title": "Resource Generator",
  "type": "object",
  "properties": {
    "projectName": {
      "type": "string",
      "description": "Application name in your workspace, e.g. 'platform'",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What is app name?",
      "x-dropdown": "projects",
      "minLength": 1
    },
    "entityName": {
      "type": "string",
      "description": "Entity that you want to generate e.g. 'user-profile'",
      "x-prompt": "What is entity name for generation? entity names should be with dashes, e.g. user-profile",
      "minLength": 1
    },
    "basePath": {
      "type": "string",
      "description": "App base path name",
      "x-prompt": "What is app base path name?  e.g. api/platform",
      "minLength": 1
    },
    "groupName": {
      "type": "string",
      "description": "Group name is folder name that you want to generate your entity in it e.g. 'users'",
      "x-prompt": "What is group name for generation? It's folder name under the base folder",
      "minLength": 1
    },
    "tenantBaseEntity": {
      "type": "boolean",
      "default": true,
      "description": "It's indicating that this entity is tenant base entity or not, so we can choose the appropriate base entity for generation",
      "x-prompt": "Is this a tenant base entity? (y/n)"
    },
    "generateRepository": {
      "type": "boolean",
      "default": true,
      "description": "It's indicating that you want to generate repository for entity or not, by default it's true because it's easier to do all in once usually",
      "x-prompt": "Do you want to generate repository for entity? (y/n)"
    },
    "generateService": {
      "type": "boolean",
      "default": true,
      "description": "It's indicating that you want to generate service or not, by default it's true because it's easier to do all in once usually",
      "x-prompt": "Do you want to generate service for entity? (y/n)"
    },
    "generateController": {
      "type": "boolean",
      "default": true,
      "description": "It's indicating that you want to generate controller or not, by default it's true because it's easier to do all in once usually",
      "x-prompt": "Do you want to generate controller for entity? (y/n)"
    },
    "entityIncludesIdField": {
      "type": "boolean",
      "default": true,
      "description": "It's indicating that you want to include 'id' field or not, by default it's true",
      "x-prompt": "Do you want to include 'id' field for entity? (y/n)"
    },
    "entityIncludesVersionField": {
      "type": "boolean",
      "default": true,
      "description": "It's indicating that you want to include 'version' field or not, by default it's true",
      "x-prompt": "Do you want to include 'version' field for entity? (y/n)"
    },
    "lintCommandName": {
      "type": "string",
      "default": "lint",
      "description": "The name of lint nx command in project.json"
    }
  },
  "required": [
    "projectName",
    "entityName",
    "tenantBaseEntity",
    "groupName",
    "generateRepository",
    "generateService",
    "generateController"
  ]
}