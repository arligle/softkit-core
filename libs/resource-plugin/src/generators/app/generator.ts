import { generateFiles, joinPathFragments, Tree, updateJson } from '@nx/devkit';
import { applicationGenerator } from '@nx/nest';

import { AppGeneratorSchema } from './schema';
import { paramCase, pascalCase, snakeCase } from 'change-case';
import i18nGenerator from '../i18n/generator';

export async function appGenerator(tree: Tree, options: AppGeneratorSchema) {
  options.name = paramCase(options.name);
  const appRoot = `apps/${options.name}`;
  const srcFolder = joinPathFragments(options.name, 'src');

  await applicationGenerator(tree, options);

  const allFilesInSrc = tree
    .listChanges()
    .filter((c) => c.path.includes(srcFolder));

  for (const f of allFilesInSrc) {
    tree.delete(f.path);
  }

  const generatorFolder = joinPathFragments(__dirname, './files');
  generateFiles(tree, generatorFolder, appRoot, {
    ...options,
    snakeCase,
    pascalCase,
  });

  if (!options.db) {
    const dbFiles = tree
      .listChanges()
      .filter(
        (c) => c.path.includes('/db/') || c.path.includes('/repositories/'),
      );

    for (const c of dbFiles) tree.delete(c.path);
  }

  if (options.i18n) {
    await i18nGenerator(tree, {
      ...options,
      buildable: true,
      baseFolder: 'apps',
    });
  }

  updateJson(
    tree,
    joinPathFragments(options.name, 'project.json'),
    (prjJson) => {
      // if scripts is undefined, set it to an empty object
      prjJson.targets.build.options.assets = [
        ...prjJson.targets.build.options.assets,
        {
          glob: '**/*',
          input: `apps/${options.name}/src/app/assets`,
          output: 'assets',
        },
      ];
      prjJson.targets.build.options.tsPlugins = [
        ...(prjJson.targets.build.options.tsPlugins || []),
        {
          name: '@nestjs/swagger/plugin',
          options: {
            dtoFileNameSuffix: [
              options.db ? '.entity.ts' : '',
              '.dto.ts',
            ].filter((value) => value !== ''),
            controllerFileNameSuffix: ['.controller.ts'],
            classValidatorShim: true,
            dtoKeyOfComment: 'description',
            controllerKeyOfComment: 'description',
            introspectComments: true,
          },
        },
      ];

      return prjJson;
    },
  );
}

export default appGenerator;
