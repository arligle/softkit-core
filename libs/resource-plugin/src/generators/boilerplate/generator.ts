import { Tree, logger, updateJson } from '@nx/devkit';

import { BoilerplateGeneratorSchema } from './schema';
import { cloneRepo } from '../common/clone-repo';
import { flushChanges } from 'nx/src/generators/tree';

async function updatePackageJson(tree: Tree, scopeName: string) {
  const packageJsonPath = 'package.json';
  logger.info(`Updating package.json at ${packageJsonPath}`);

  updateJson(tree, packageJsonPath, (json) => {
    json.name = `@${scopeName}/source`;
    return json;
  });
}

export async function platformGenerator(
  _: Tree,
  options: BoilerplateGeneratorSchema,
) {
  const root = options.appFolder;
  options.companyName = options.companyName.toLowerCase().trim();

  logger.info(`Cloning repository into ${root}...`);
  const fsTree = await cloneRepo(root, options.tag, options.repository);

  await updatePackageJson(fsTree, options.companyName);

  flushChanges(root, fsTree.listChanges());
}

export default platformGenerator;
