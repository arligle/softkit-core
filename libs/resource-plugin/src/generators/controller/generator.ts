import {
  generateFiles,
  joinPathFragments,
  readProjectConfiguration,
  Tree,
} from '@nx/devkit';
import { ControllerGeneratorSchema } from './schema';
import { capitalCase, paramCase, pascalCase, snakeCase } from 'change-case';
import { EOL } from 'node:os';

export async function controllerGenerator(
  tree: Tree,
  options: ControllerGeneratorSchema,
) {
  const appRoot = readProjectConfiguration(tree, options.projectName).root;

  const srcFolder = joinPathFragments(__dirname, './files');
  generateFiles(tree, srcFolder, appRoot, {
    ...options,
    snakeCase,
    pascalCase,
    paramCase,
    capitalCase,
  });

  const controllersFolder = joinPathFragments(appRoot, 'src/app/controllers');
  const groupName = options.groupName || '';

  const exportPathForIndex = joinPathFragments(
    groupName,
    `${options.controllerName}.controller`,
  );

  const indexFilePath = joinPathFragments(controllersFolder, `index.ts`);

  const contents = tree.exists(indexFilePath)
    ? tree.read(indexFilePath).toString()
    : '';
  const newContents = `${contents}${EOL}export * from './${exportPathForIndex}';`;
  tree.write(indexFilePath, newContents);
  //   todo run eslint --fix
}

export default controllerGenerator;
