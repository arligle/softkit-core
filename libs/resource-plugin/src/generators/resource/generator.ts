import {
  generateFiles,
  joinPathFragments,
  readProjectConfiguration,
  Tree,
} from '@nx/devkit';
import { EOL } from 'node:os';
import { ResourceGeneratorSchema } from './schema';
import { pascalCase, snakeCase } from 'change-case';
import {} from 'node:child_process';
import repositoryGenerator from '../repository/generator';
import { RepositoryGeneratorSchema } from '../repository/schema';
import serviceGenerator from '../service/generator';
import { ServiceGeneratorSchema } from '../service/schema';
import controllerGenerator from '../controller/generator';
import { ControllerGeneratorSchema } from '../controller/schema';

export async function resourceGenerator(
  tree: Tree,
  options: ResourceGeneratorSchema,
) {
  const appRoot = readProjectConfiguration(tree, options.projectName).root;

  const srcFolder = joinPathFragments(__dirname, './files');
  generateFiles(tree, srcFolder, appRoot, {
    ...options,
    snakeCase,
    pascalCase,
  });

  const entitiesFolder = joinPathFragments(
    appRoot,
    'src/app/database/entities',
  );
  const groupName = options.groupName || '';

  const exportPathForIndex = joinPathFragments(
    groupName,
    `${options.entityName}.entity`,
  );

  const indexFilePath = joinPathFragments(entitiesFolder, `index.ts`);

  const contents = tree.exists(indexFilePath)
    ? tree.read(indexFilePath).toString()
    : '';
  const newContents = `${contents}${EOL}export * from './${exportPathForIndex}';`;
  tree.write(indexFilePath, newContents);
  //   todo run eslint --fix

  if (options.generateRepository) {
    const repositoryOptions = {
      ...options,
      tenantBaseRepository: options.tenantBaseEntity,
      repositoryName: `${options.entityName}`,
    } satisfies RepositoryGeneratorSchema;
    await repositoryGenerator(tree, repositoryOptions);
  }

  if (options.generateService) {
    const repositoryOptions = {
      ...options,
      tenantBaseService: options.tenantBaseEntity,
      repositoryName: `${options.entityName}`,
      serviceName: options.entityName,
    } satisfies ServiceGeneratorSchema;
    await serviceGenerator(tree, repositoryOptions);
  }

  if (options.generateController) {
    const repositoryOptions = {
      ...options,
      serviceName: options.entityName,
      controllerName: options.entityName,
    } satisfies ControllerGeneratorSchema;
    await controllerGenerator(tree, repositoryOptions);
  }
}

export default resourceGenerator;
