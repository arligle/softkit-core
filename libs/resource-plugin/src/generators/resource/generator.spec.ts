import { Tree, readProjectConfiguration } from '@nx/devkit';

import { resourceGenerator } from './generator';
import { ResourceGeneratorSchema } from './schema';
import { createTreeWithNestApplication } from '@nx/nest/src/generators/utils/testing';
import { EOL } from 'node:os';

describe('resource generator', () => {
  let tree: Tree;
  const options: ResourceGeneratorSchema = {
    projectName: 'test-project',
    entityName: 'test-entity',
    groupName: 'test-group',
    basePath: 'api/platform',
    tenantBaseEntity: true,
    generateRepository: true,
    generateService: true,
    generateController: true,
    entityIncludesIdField: true,
    entityIncludesVersionField: true,
  };

  beforeEach(() => {
    tree = createTreeWithNestApplication(options.projectName);
  });

  it('should creat files for all embedded resources and appropriate folders', async () => {
    const changesBeforeGeneratorRun = [...tree.listChanges()];
    await resourceGenerator(tree, options);
    const config = readProjectConfiguration(tree, options.projectName);
    expect(config).toBeDefined();

    const changesAfterGenerator = tree
      .listChanges()
      .slice(changesBeforeGeneratorRun.length);

    // this generator should create 2 files
    expect(changesAfterGenerator.length).toBe(12);

    const indexTsFile = changesAfterGenerator.find((change) => {
      return change.path.includes('index');
    });

    expect(indexTsFile).toBeDefined();
    expect(indexTsFile.content.toString()).toContain(options.entityName);
    // group name should be a folder in the path
    expect(indexTsFile.content.toString()).toContain(`/${options.groupName}/`);

    const entityFile = changesAfterGenerator.find((change) => {
      return change.path.includes(options.entityName);
    });

    expect(entityFile.path).toContain(`/${options.groupName}/`);
    expect(entityFile).toBeDefined();

    await resourceGenerator(tree, {
      ...options,
      entityName: 'second-entity',
    });

    const indexFile = tree.listChanges().find((change) => {
      return change.path.includes('index.ts');
    });

    expect(indexFile.content.toString().split(EOL).length).toBe(3);
  });
});
