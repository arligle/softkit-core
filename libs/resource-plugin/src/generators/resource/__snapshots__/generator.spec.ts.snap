// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resource generator should run successfully 1`] = `
[
  {
    "content": {
      "data": [
        105,
        109,
        112,
        111,
        114,
        116,
        32,
        123,
        10,
        32,
        32,
        69,
        110,
        116,
        105,
        116,
        121,
        44,
        10,
        32,
        32,
        80,
        114,
        105,
        109,
        97,
        114,
        121,
        71,
        101,
        110,
        101,
        114,
        97,
        116,
        101,
        100,
        67,
        111,
        108,
        117,
        109,
        110,
        44,
        10,
        125,
        32,
        102,
        114,
        111,
        109,
        32,
        39,
        116,
        121,
        112,
        101,
        111,
        114,
        109,
        39,
        59,
        10,
        10,
        105,
        109,
        112,
        111,
        114,
        116,
        32,
        123,
        10,
        10,
        32,
        32,
        32,
        32,
        66,
        97,
        115,
        101,
        84,
        101,
        110,
        97,
        110,
        116,
        69,
        110,
        116,
        105,
        116,
        121,
        72,
        101,
        108,
        112,
        101,
        114,
        32,
        97,
        115,
        32,
        66,
        97,
        115,
        101,
        69,
        110,
        116,
        105,
        116,
        121,
        10,
        10,
        125,
        32,
        102,
        114,
        111,
        109,
        32,
        39,
        64,
        115,
        97,
        97,
        115,
        45,
        98,
        117,
        105,
        108,
        100,
        107,
        105,
        116,
        47,
        116,
        121,
        112,
        101,
        111,
        114,
        109,
        39,
        59,
        10,
        10,
        64,
        69,
        110,
        116,
        105,
        116,
        121,
        40,
        39,
        116,
        101,
        115,
        116,
        95,
        101,
        110,
        116,
        105,
        116,
        121,
        39,
        41,
        10,
        101,
        120,
        112,
        111,
        114,
        116,
        32,
        99,
        108,
        97,
        115,
        115,
        32,
        84,
        101,
        115,
        116,
        69,
        110,
        116,
        105,
        116,
        121,
        32,
        101,
        120,
        116,
        101,
        110,
        100,
        115,
        32,
        66,
        97,
        115,
        101,
        69,
        110,
        116,
        105,
        116,
        121,
        32,
        123,
        10,
        32,
        32,
        64,
        80,
        114,
        105,
        109,
        97,
        114,
        121,
        71,
        101,
        110,
        101,
        114,
        97,
        116,
        101,
        100,
        67,
        111,
        108,
        117,
        109,
        110,
        40,
        39,
        117,
        117,
        105,
        100,
        39,
        41,
        10,
        32,
        32,
        105,
        100,
        33,
        58,
        32,
        115,
        116,
        114,
        105,
        110,
        103,
        59,
        10,
        125,
        10,
      ],
      "type": "Buffer",
    },
    "options": undefined,
    "path": "apps/test-project/src/app/database/entities/test-group/test-entity.entity.ts",
    "type": "CREATE",
  },
  {
    "content": {
      "data": [
        10,
        101,
        120,
        112,
        111,
        114,
        116,
        32,
        42,
        32,
        102,
        114,
        111,
        109,
        32,
        39,
        46,
        47,
        116,
        101,
        115,
        116,
        45,
        103,
        114,
        111,
        117,
        112,
        47,
        116,
        101,
        115,
        116,
        45,
        101,
        110,
        116,
        105,
        116,
        121,
        46,
        101,
        110,
        116,
        105,
        116,
        121,
        39,
        59,
      ],
      "type": "Buffer",
    },
    "options": undefined,
    "path": "apps/test-project/src/app/database/entities/index.ts",
    "type": "CREATE",
  },
]
`;
